### PR TITLE
drivers/sdcard_spi: provide mtd interface for sdcard

### PR DESCRIPTION
**Pull request by MichelRottleuthner**

### CODE REVIEW

**Comment by vincent-d** [comment]
Nice one!

I took a quick look, and I'm not sure to understand why you need an `auto_init` and the `mtd_sdcard_t mtd_sdcard_devs[SDCARD_SPI_NUM]` array.

IMO, the mtd device should be initialized only when mounting the file system.

The app is then in charge of allocating the mtd device descriptor. That's at least how we did with `mtd_spi_nor`.

Maybe @gebart has an opinion on that?

I'll try to test if I have some time and I can find a suitable board ;)

**Comment by gebart** [comment]
I agree, I don't see the need for an auto init for mtd_sdcard. Potentially there might be a use for a general MTD auto init which would run init on all mtd devices defined at compile time, but the same might also be placed in the application or board init instead.

**Comment by MichelRottleuthner** [comment]
You are right, it's a leftover of my experiments with devfs. I was trying to make the test app for FatFs more generic by using autoinit/devfs but then decided against it. 

>I'll try to test if I have some time and I can find a suitable board ;)

Every board with spi should do if you mange to get the sd-card wired up ;)

File: drivers/sdcard_spi/include/sdcard_spi_params.h [hunk 1]
```diff
@@ -69,6 +69,8 @@ static const  sdcard_spi_params_t sdcard_spi_params[] = {
 };
 /** @} */
 
+#define SDCARD_SPI_NUM (sizeof(sdcard_spi_params) / sizeof(sdcard_spi_params[0]))
+
```
Comment (vincent-d) [inline drivers/sdcard_spi/include/sdcard_spi_params.h]: I guess this can be removed as well

File: drivers/mtd_sdcard/mtd_sdcard.c [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+static int mtd_sdcard_erase(mtd_dev_t *dev,
+                 uint32_t addr,
+                 uint32_t size)
+{
+    DEBUG("mtd_sdcard_erase: addr:%lu size:%lu\n", addr, size);
+    (void)dev;
+    (void)addr;
+    (void)size;
+
+    #ifdef MTD_SDCARD_SKIP_ERASE
```
Comment (vincent-d) [inline drivers/mtd_sdcard/mtd_sdcard.c]: indentation, here and below

**Comment by MichelRottleuthner** [comment]
@vincent-d thanks for testing! 
>(...)underlying driver which hangs sometimes

can you share more details on this, like what exactly hangs, what card you are using, your setup, etc. This would greatly help with improving that driver.

**Comment by vincent-d** [comment]
> like what exactly hangs, what card you are using, your setup, etc.

I had to disable optimizations to go through `mtd_init`. Then the `mtd_read` sometimes return an error.

I used a 4G Kingston Micro SDHC card, with a shield having a sd card slot. My board is a nucleo144-f413.

I had no time to investigate further.

File: drivers/mtd_sdcard/Makefile [hunk 3]
```diff
@@ -0,0 +1,3 @@
+MODULE = mtd_sdcard
```
Comment (vincent-d) [inline drivers/mtd_sdcard/Makefile]: This is not needed as the directory name is used by default

**Comment by MichelRottleuthner** [comment]
@vincent-d did you use arbitrary sizes for r/w? In the current state r/w only supports multiples of 512 bytes so no extra buffer is needed in the mtd layer. 
Unfortunately I couldn't reproduce faulty behaviour when using SD-cards over mtd (tested multiple cards on multiple platforms). I suspect the problem has to do with timeouts of sdcard_spi that are not sufficient for the fast nucleo144-f413. You could test if increasing the retry counts in [sdcard_spi_internal.h](https://github.com/RIOT-OS/RIOT/blob/master/drivers/sdcard_spi/include/sdcard_spi_internal.h#L131) improves it. I'll try to find a nucleo144-f413 for further testing.

**Comment by vincent-d** [comment]
I gave a new try and could fix my issues by changing the `XTIMER_BACKOFF` to 10, otherwise it hangs in the `SD_INIT_SPI_POWER_SEQ` step of the init function.

File: drivers/include/mtd_sdcard.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+
+/**
+ * @brief sdcards handle sector erase internally so it's possible to directly
+ *        write to the card without erasing the sector first.
+ *        Attention: an erase call will therefore NOT touch the content,
+ *                   so disable this feature to ensure overriding the data.
+ */
+#ifndef MTD_SDCARD_SKIP_ERASE
+#define MTD_SDCARD_SKIP_ERASE
```
Comment (vincent-d) [inline drivers/include/mtd_sdcard.h]: I'd rather define `MTD_SDCARD_SKIP_ERASE` to 0 or 1 and use `#if MTD_SDCARD_SKIP_ERASE == 1` in the code. As is, it cannot be disabled.

File: drivers/include/mtd_sdcard.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ * @author      Michel Rottleuthner <michel.rottleuthner@haw-hamburg.de>
+ */
+
+#ifndef MTD_SDCARD_H_
+#define MTD_SDCARD_H_
+
+#include <stdint.h>
+
+#include "sdcard_spi.h"
+#include "sdcard_spi_params.h"
```
Comment (vincent-d) [inline drivers/include/mtd_sdcard.h]: This include seems not needed.

Comment (MichelRottleuthner) [inline drivers/include/mtd_sdcard.h]: fixed

File: drivers/include/mtd_sdcard.h (hunk 4 above)
Comment (MichelRottleuthner) [inline drivers/include/mtd_sdcard.h]: fixed

**Comment by MichelRottleuthner** [comment]
thanks @vincent-d for testing this again. Ignore the two commits that I pushed accidentally and reverted, the last one addresses your comments.

File: drivers/mtd_sdcard/mtd_sdcard.c [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    .erase = mtd_sdcard_erase,
+    .power = mtd_sdcard_power,
+};
+
+static int mtd_sdcard_init(mtd_dev_t *dev)
+{
+    DEBUG("mtd_sdcard_init\n");
+    mtd_sdcard_t *mtd_sd = (mtd_sdcard_t*)dev;
+    if ( (mtd_sd->sd_card->init_done == true) ||
+         (sdcard_spi_init(mtd_sd->sd_card, mtd_sd->params) == 0) ) {
```
Comment (vincent-d) [inline drivers/mtd_sdcard/mtd_sdcard.c]: Extra spaces

File: drivers/mtd_sdcard/mtd_sdcard.c [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+static int mtd_sdcard_power(mtd_dev_t *dev, enum mtd_power_state power)
+{
+    (void)dev;
+    (void)power;
+
+    /* TODO: implement power down of sdcard in sdcard_spi
+    (make use of sdcard_spi_params_t.power pin) */
+    return -ENOTSUP; /**< currently not supported */
```
Comment (vincent-d) [inline drivers/mtd_sdcard/mtd_sdcard.c]: No need for doxygen comments

File: drivers/mtd_sdcard/mtd_sdcard.c [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ * @author      Michel Rottleuthner <michel.rottleuthner@haw-hamburg.de>
+ *
+ * @}
+ */
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+#include "mtd.h"
+#include "mtd_sdcard.h"
+#include "sdcard_spi.h"
+#include "sdcard_spi_internal.h"
```
Comment (vincent-d) [inline drivers/mtd_sdcard/mtd_sdcard.c]: Is this include needed?

Comment (MichelRottleuthner) [inline drivers/mtd_sdcard/mtd_sdcard.c]: yes, `sdcard_spi_get_sector_count` is declared there.

File: drivers/mtd_sdcard/mtd_sdcard.c (hunk 6 above)
Comment (MichelRottleuthner) [inline drivers/mtd_sdcard/mtd_sdcard.c]: fixed

File: drivers/mtd_sdcard/mtd_sdcard.c (hunk 7 above)
Comment (MichelRottleuthner) [inline drivers/mtd_sdcard/mtd_sdcard.c]: fixed

**Comment by MichelRottleuthner** [comment]
addressed last comments and squashed

**Comment by vincent-d** [comment]
ACK

**Pull request merged**